text/0000-inline-semantic.md

```diff
@@ -0,0 +1,993 @@
+
+## Semantic-inlining
+
+The reason a `my_unwrap!` macro works is because it copy-and-paste the entire content of its macro
+definition everytime it is used.
+
+```rust
+println!("args[1] = {}", my_unwrap!(args().nth(1)));
+println!("args[2] = {}", my_unwrap!(args().nth(2)));
+...
+
+// is equivalent to:
+
+println!("args[1] = {}", match args().nth(1) {
+    Some(t) => t,
+    None => panic!("nothing to see at {}:{}:{}, move along", file!(), line!(), column!()),
+});
+println!("args[2] = {}", match args().nth(2) {
+    Some(t) => t,
+    None => panic!("nothing to see at {}:{}:{}, move along", file!(), line!(), column!()),
+});
+...
+```
+
+What if we allow *normal functions* be able to copy-and-paste its content as well? This is called
+**[inlining]**, which Rust and many other languages have supported this as an optimization step. If
+the function is inlined, the compiler can also tell the inlined function the location it is inlined,
+and thus able to report this to the user.
+
+Rust allows developers to use the `#[inline]` attribute to *hint* the optimizer that a function
+should be inlined. However, if we want the precise caller location, a hint is not enough, it needs
+to be a requirement. Therefore, the `#[inline(semantic)]` attribute is introduced.
+
+```rust
+#![feature(inline_semantic)]
+
+#[inline(semantic)] // <-- new
+pub fn my_unwrap<T>(input: Option<T>) -> T {
+    match input {
+        Some(t) => t,
+        None => panic!("nothing to see here, move along"),
+    }
+}
+
+println!("args[1] = {}", my_unwrap(args().nth(1)));
+println!("args[2] = {}", my_unwrap(args().nth(2)));
+
+// almost equivalent to:
+
+println!("args[1] = {}", match args().nth(1) {
+    Some(t) => t,
+    None => panic!("nothing to see here, move along"),
+});
+println!("args[2] = {}", match args().nth(2) {
+    Some(t) => t,
+    None => panic!("nothing to see here, move along"),
+});
+```
+
+If you try this code above, you will find that the panic still occurs inside `my_unwrap`, not at the
+caller's position. `#[inline(semantic)]` is still only a very-insistent inliner, so its behavior is
+still like a normal function — the syntactic location of the `panic!` is still inside the `my_wrap`.
+The Rust language provides a different way to obtain the caller location after the function is
+inlined.
+
+## Caller location
+
+The core crate provides three magic constants `core::caller::{FILE, LINE, COLUMN}` which resolves to
+the caller's location one the function is inlined.
+
+```rust
+#![feature(inline_semantic, caller_location)]
+
+extern crate core;
+
+#[inline_semantic]
+pub fn get_caller_loc() -> (&'static str, u32) {
+    (core::caller::FILE, core::caller::LINE)
+}
+
+assert_eq!(get_caller_loc(), (file!(), line!()));
+assert_eq!(get_caller_loc(), (file!(), line!()));
+assert_eq!(get_caller_loc(), (file!(), line!()));
+```
+
+There is also a `caller_location!()` macro to return all three information as a single tuple.
```

Comment:
This doesn't need to be a macro, right? Could it be a semantically inlined function instead?